lib/Ui/Tree.elm

```diff
@@ -0,0 +1,178 @@
+        << Tree.restructure identity
+            (\n c ->
+                case c of
+                    [] ->
+                        tree (Leaf n) []
+
+                    children ->
+                        tree (Collapsed n) children
+            )
+
+
+type Node a
+    = Collapsed a
+    | Expanded a
+    | Leaf a
+
+
+unwrapNode : Node a -> a
+unwrapNode n =
+    case n of
+        Collapsed a ->
+            a
+
+        Expanded a ->
+            a
+
+        Leaf a ->
+            a
+
+
+isExpanded : Node a -> Bool
+isExpanded n =
+    case n of
+        Collapsed _ ->
+            False
+
+        Expanded _ ->
+            True
+
+        Leaf _ ->
+            False
+
+
+isCollapsed : Node a -> Bool
+isCollapsed =
+    not << isExpanded
+
+
+type Msg a
+    = Expand (Node a)
+    | Collapse (Node a)
+    | Select (Node a)
+
+
+
+-- type Msg a comparable = Collapse comparable
+--     | Expand comparable
+--     | Select comparable
+--     | Up (Zipper a)
+--     | Down (Zipper a)
+
+
+view : { showLabel : a -> String, liftMsg : Msg a -> msg } -> Model a -> Styled.Html msg
+view { showLabel, liftMsg } model =
+    let
+        viewChildren level =
+            Styled.ul [ css [ Css.marginLeft <| (Css.px <| toFloat (level * 35)) ] ] << List.map (viewTree level) << Tree.children
+
+        viewTree : Int -> Tree (Node a) -> Styled.Html msg
+        viewTree level subtree =
+            let
+                node =
+                    Tree.label subtree
+
+                labelStyle =
+                    css <|
+                        [ Css.cursor Css.pointer
+                        , Css.padding2 (Css.px 15) (Css.px 15)
+                        , Css.borderRadius (Css.px 5)
+                        , Css.hover [ Css.backgroundColor <| toCssColor Ui.Palette.grey100 ]
+                        ] ++ if Zipper.label model == node then
+                                [ Css.backgroundColor <| toCssColor Ui.Palette.grey200 ]
+                            else []
+
+
+                chevronStyle =
+                    css
+                        [ Css.height (Css.px 10)
+                        , Css.display Css.inlineBlock
+                        , Css.marginRight (Css.px 10)
+                        ]
+            in
+            case node of
+                Collapsed _ ->
+                    Styled.li [  ]
+                        [ Styled.div [ labelStyle, StyledEvents.onClick <| liftMsg (Expand node) ]
+                            [ Styled.span
+                                [ chevronStyle
+                                ]
+                                [ Icon.view Icon.chevronRight ]
+                            , Styled.text <| (showLabel <| unwrapNode node) ++ " - level " ++ String.fromInt level
+                            ]
+                        ]
+
+                Expanded _ ->
+                    Styled.li [ ]
+                        [ Styled.div [ labelStyle,  StyledEvents.onClick <| liftMsg (Collapse node) ]
```

Comment:
I thought we should allow selecting non-leaf categories

Reply:
We do, it selects as well